srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,1606 @@
+		/// successful or not (third element).
+		VoterReported(AccountId, AccountId, bool),
+	}
+);
+
+impl<T: Trait> Module<T> {
+	/// Check if `who` is a candidate. It returns the insert index if the element does not exists as
+	/// an error.
+	///
+	/// State: O(LogN) given N candidates.
+	fn is_candidate(who: &T::AccountId) -> Result<(), usize> {
+		Self::candidates().binary_search(who).map(|_| ())
+	}
+
+	/// Check if `who` is a voter. It may or may not be a _current_ one.
+	///
+	/// State: O(1).
+	fn is_voter(who: &T::AccountId) -> bool {
+		<StakeOf<T>>::exists(who)
+	}
+
+	/// Check if `who` is currently an active member.
+	///
+	/// Limited number of members. Binary search. Constant time factor. O(1)
+	fn is_member(who: &T::AccountId) -> bool {
+		Self::members().binary_search(who).is_ok()
+	}
+
+	/// Check if `who` is a defunct voter.
+	///
+	/// Note that false is returned if `who` is not a voter at all.
+	///
+	/// O(NLogM) with M candidates and `who` having voted for `N` of them.
+	fn is_defunct_voter(who: &T::AccountId) -> bool {
+		if Self::is_voter(who) {
+			Self::votes_of(who)
+				.iter()
+				.all(|v| !Self::is_member(v) && !Self::is_candidate(v).is_ok())
+		} else {
+			false
+		}
+	}
+
+	/// Remove a certain someone as a voter.
+	///
+	/// This will clean always clean the storage associated with the voter, and remove the balance
+	/// lock. Optionally, it would also return the reserved voting bond if indicated by `unreserve`.
+	fn do_remove_voter(who: &T::AccountId, unreserve: bool) {
+		// remove storage and lock.
+		<VotesOf<T>>::remove(who);
+		<StakeOf<T>>::remove(who);
+		T::Currency::remove_lock(MODULE_ID, who);
+
+		if unreserve {
+			T::Currency::unreserve(who, T::VotingBond::get());
+		}
+	}
+
+	/// The locked stake of a voter.
+	fn locked_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		Self::stake_of(who)
+	}
+
+	/// Check there's nothing to do this block.
+	///
+	/// Runs phragmen election and cleans all the previous candidate state. The voter state is NOT
+	/// cleaned and voters must themselves submit a transaction to retract.
+	fn end_block(block_number: T::BlockNumber) -> dispatch::Result {
+		if (block_number % Self::term_duration()).is_zero() {
+			Self::do_phragmen();
+		}
+		Ok(())
+	}
+
+	/// Run the phragmen election with all required side processes and state updates.
+	///
+	/// Calls the appropriate `ChangeMembers` function variant internally.
+	///
+	/// # <weight>
+	/// #### State
+	/// Reads: O(C + V*E) where C = candidates, V voters and E votes per voter exits.
+	/// Writes: O(M + R) with M desired members and R runners_up.
+	/// # </weight>
+	fn do_phragmen() {
+		let desired_seats = Self::desired_members() as usize;
+		let desired_runners_up = Self::desired_runners_up() as usize;
+		let num_to_elect = desired_runners_up + desired_seats;
+
+		let mut candidates = Self::candidates();
+		// candidates who exactly called `submit_candidacy`. Only these folks are at the risk of
```

Comment:
```suggestion
		// candidates who explicitly called `submit_candidacy`. Only these folks are at the risk of
```

Reply:
(?)

Reply:
The terminology might be strange once seen isolated, but here's the full story: 

As requested and discussed, current members and runners up are kept as candidates for future rounds. To do this, we don't duplicate them into `Candidate<T>` but (the way I named it) _implicitely_ add them as candidates. As opposed to this, some ppl will always be candidates who made it in by `submit_candidacy()`. I named this _explicit candidates_. Both namings can work.